src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs

```diff
@@ -320,5 +320,165 @@ public void TestUdfWithReturnAsRowType()
+                {
+                    new StructField("col1", new StringType()),
+                    new StructField("col2", subSchema1),
+                });
+                var schema = new StructType(new[]
+                {
+                    new StructField("col1", new IntegerType()),
+                    new StructField("col2", subSchema1),
+                    new StructField("col3", subSchema2)
+                });
+
+                _df.CreateOrReplaceTempView("people");
+
+                _spark.Udf().Register<string>(
+                    "udf3",
+                    str => new GenericRow(
+                        new object[]
+                        {
+                            1,
+                            new GenericRow(new object[] { 1 }),
+                            new GenericRow(new object[]
+                                {
+                                    "abc",
+                                    new GenericRow(new object[] { 10 })
+                                })
+                        }),
+                    schema);
+
+                Row[] rows =
+                    _spark.Sql("SELECT udf3(name) AS col FROM people")
+                    .Collect()
+                    .ToArray();
+                Assert.Equal(3, rows.Length);
+
+                foreach (Row row in rows)
+                {
+                    Assert.Equal(1, row.Size());
+                    Row outerCol = row.GetAs<Row>("col");
+                    Assert.Equal(3, outerCol.Size());
+                    Assert.Equal(1, outerCol.GetAs<int>("col1"));
+                    Assert.Equal(
+                        new Row(new object[] { 1 }, subSchema1),
+                        outerCol.GetAs<Row>("col2"));
+                    Assert.Equal(
+                        new Row(
+                            new object[] { "abc", new Row(new object[] { 10 }, subSchema1) },
+                            subSchema2),
+                        outerCol.GetAs<Row>("col3"));
+                }
+            }
+
+            // Chained UDFs.
+            {
+                var schema = new StructType(new[]
+                {
+                    new StructField("col1", new IntegerType()),
+                    new StructField("col2", new StringType())
+                });
+
+                _df.CreateOrReplaceTempView("people");
+
+                _spark.Udf().Register<string>(
+                    "udf4",
+                    str => new GenericRow(new object[] { 1, str }),
+                    schema);
+
+                _spark.Udf().Register<Row, string>(
+                    "udf5",
+                    row => row.GetAs<string>(1));
+
+                Row[] rows =
+                    _spark.Sql("SELECT udf5(udf4(name)) FROM people")
+                    .Collect()
+                    .ToArray();
+                Assert.Equal(3, rows.Length);
+
+                var expected = new[] { "Michael", "Andy", "Justin" };

```

Comment:
nit: Probably existing code, but it should be `var expected = new string[] { ... }`;

Reply:
I have made these changes in the latest commits